src/script/lua_api/l_ipc.cpp

```diff
@@ -0,0 +1,138 @@
+// Minetest
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+#include "lua_api/l_ipc.h"
+#include "lua_api/l_internal.h"
+#include "common/c_packer.h"
+#include "server.h"
+#include "debug.h"
+#include <chrono>
+
+typedef std::shared_lock<std::shared_mutex> SharedReadLock;
+typedef std::unique_lock<std::shared_mutex> SharedWriteLock;
+
+static inline auto read_pv(lua_State *L, int idx)
+{
+	std::unique_ptr<PackedValue> ret;
+	if (!lua_isnil(L, idx)) {
+		ret.reset(script_pack(L, idx));
+		if (ret->contains_userdata)
+			throw LuaError("Userdata not allowed");
+	}
+	return ret;
+}
+
+int ModApiIPC::l_ipc_get(lua_State *L)
+{
+	auto *store = getGameDef(L)->getModIPCStore();
+
+	auto key = readParam<std::string>(L, 1);
+
+	{
+		SharedReadLock autolock(store->mutex);
+		auto it = store->map.find(key);
+		if (it == store->map.end())
+			lua_pushnil(L);
+		else
+			script_unpack(L, it->second.get());
+	}
+	return 1;
+}
+
+int ModApiIPC::l_ipc_set(lua_State *L)
+{
+	auto *store = getGameDef(L)->getModIPCStore();
+
+	auto key = readParam<std::string>(L, 1);
+
+	luaL_checkany(L, 2);
+	auto pv = read_pv(L, 2);
+
+	{
+		SharedWriteLock autolock(store->mutex);
+		if (pv)
+			store->map[key] = std::move(pv);
+		else
+			store->map.erase(key); // delete the map value for nil
+		store->signal();
+	}
+	return 0;
+}
+
+int ModApiIPC::l_ipc_cas(lua_State *L)
+{
+	auto *store = getGameDef(L)->getModIPCStore();
+
+	auto key = readParam<std::string>(L, 1);
+
+	luaL_checkany(L, 2);
+	const int idx_old = 2;
+
+	luaL_checkany(L, 3);
+	auto pv_new = read_pv(L, 3);
+
+	bool ok = false;
+	{
+		SharedWriteLock autolock(store->mutex);
+		// unpack and compare old value
+		auto it = store->map.find(key);
+		if (it == store->map.end()) {
+			ok = lua_isnil(L, idx_old);
+		} else {
+			script_unpack(L, it->second.get());
+			ok = lua_equal(L, idx_old, -1);
+			lua_pop(L, 1);
```

Comment:
out of curiosity, why can't the packed values be compared directly without having to first unpack them into Lua values?

Reply:
It is possible to pack tables while "preserving" their metatables (particularly the `__eq` metamethod), so values _potentially_ need to be unpacked into Lua to run the `__eq` metamethod.

Reply:
is that something that *should* be supported for a CAS operation? It feels a bit low level for that.

Reply:
it's not even about metatables but doing so would require me to implement type coercion Lua does for comparisons.
I think it does not matter in practice because CAS barely makes sense with complex values and unpacking *simple* values is practically free.

Reply:
> type coercion Lua does for comparisons

huh, lua does no type coercion for comparisons? values need to have the same type to be equal.

---

i know this is already resolved (and i agree it is) but in case someone thinks of "optimizing" comparisons by comparing serialized values, i still wanted to add that lua uses hash maps and guarantees no order. efficient packing won't be sorting the keys. so the "same" table could be packed in two different ways depending on how it was constructed.

Reply:
I checked and you are correct, Lua doesn't do type coercion for comparison.